initial lokinet router testing

majestrate opened the pull request:
* report via rpc to oxen core connection stats on success and failure
* connect to random service node by pubkey every 5 seconds for testing

majestrate committed:
initial lokinet router testing:
* report via rpc to oxen core connection stats on success and failure
* connect to random service node by pubkey every 5 seconds for testing

tewinget commented on llarp/rpc/lokid_rpc_client.hpp:
```diff
@@ -42,6 +42,10 @@ namespace llarp
           dht::Key_t namehash,
           std::function<void(std::optional<service::EncryptedName>)> resultHandler);
 
+      /// record that if connected to a router successfully
```
typo in comment

jagerman commented:
Some suggestions from recently doing this on the storage server side:

- random testing is fairly erratic because some nodes can go a very long time without getting tested.  Rather in SS we use a shuffled list and each time we test a node we pull off the top, and regenerate the list when it empties.
- 5s seems a bit faster than necessary (and we can make it a bit slower especially if we go through the list randomly, rather than just randomly selected each time).
- oxend expects frequent retesting of failing nodes (and needs it, as it considers a node to be no longer failing if it doesn't get an update within 5 minutes).  SS's (recently reworked) logic on this is:
  - keep a list of failing nodes, along with their scheduled re-test time and number of previous failures.
  - on a test tick, remove one random node from the shuffled list (skipping any that are already on the fail list), plus nodes on the fail list due to be tested.
  - when a node fails, add it onto the failing list, with a re-test time of:
    - 10s + R from now if it was not previously failing
    - 10s*N + R from now if it was previously failing with N previous failures
    - R here is some random fuzz (so that we space out re-tests), drawn from a `inline static thread_local std::normal_distribution<float> TESTING_INTERVAL{10.0, 3.0};`
- tick the re-test queue faster (SS uses every 100ms); most of the time it will do nothing, but this gives more spreading out of test times.
- keep track of when the next "general" (i.e. random working snode) test is and, after each general test, reset it to `now + TESTING_INTERVAL(rng())` (so that we also space out random testing, but keep them at 10s apart on average).

You can probably actually lift storage server's reachability_testing.cpp/.h code more or less as is (really the only thing SS-specific in it is the `swarm` object—used to access the current list of active nodes from within storage server—and the logger).

majestrate committed:
* use weak_ptr on core rpc
* use reachability testing code lifted storage server's code

majestrate committed:
there were some unhandled edge cases in outbound_session_maker, specifically when we are not permitted to connect to a remoute but we got its rc we silently drop the error.

jagerman committed:
Switch stl mt19937_64 to CSRNG

majestrate committed:
capture by value to appease clang 11

majestrate commented on llarp/rpc/lokid_rpc_client.cpp:
```diff
@@ -190,9 +197,16 @@ namespace llarp
+            if (rid.FromHex(ed_itr->get<std::string>()) and pk.FromHex(svc_itr->get<std::string>()))
+            {
               nodeList.emplace_back(std::move(rid));
+              keymap[rid] = pk;
```
this here, move above `nodeList.emplace_back`

jagerman committed:
Fix use after move

jagerman committed:
Add logging about testing success/fail

majestrate committed:
handle case where we already have an outbound session, inform caller about it

majestrate committed:
separate white/grey list for active/decommissioned nodes.

allow sessions to decommissioned nodes but not paths.

jagerman committed:
Add warning when can't update whitelist

jagerman committed:
Fix unintentional whitelist/greylist copying

jagerman committed:
Avoid copying & keep router alive by moving shared_ptr

jagerman committed:
Don't pass last hash into UpdateServiceNodeList

It's there for polling, which we aren't doing anymore; we just got the
hash from oxend's push notification, so if it pass it in then we will
always get an "unchanged" result because we're telling oxend that we
already have the data for that hash updated.

This just drops the hash completely because we don't need it anymore.

jagerman committed:
Fix failing pks not being populated

jagerman committed:
Make format

majestrate committed:
gossip RC when we are not deregistered so we can come back when we are decommissioned

majestrate committed:
typofix

majestrate committed:
dont run router testing if we are decommissioned.

properly name function to be called LooksDecommissioned because that is different than deregistered

tewinget committed:
make outbound session if we do not have

currently creating an outbound session will cancel if we have any session
at all with the relay.  instead, only cancel if we have an outbound session
to that relay.  this is useful for reachability testing.

majestrate committed:
make windows happy by making some constexprs non static and such as windows does not LTO

majestrate commented:
ah yes, the monthly patch tuesday github outage breaks CI... amazin.

majestrate committed:
* get_failing does not need abstract router as paramter so we remove it
* add remove_node_from_failing to remove a node by pubkey from the failing set
* if a router is deregistered we remove it from the failing set so we don't retest it
* remove a router from the failing set if we get a test success
